README
libopensmskummeeplugin

libopensmskummeeplugin is an OpenSM performance manager plugin.  The plugin
receives data from the Performance Manager and logs that data to a MySQL
database.  Scripts are provided for integration to SKUMMEE.  However, these
scripts are not required to use the plugin; direct DB queries are possible.
Furthermore, a CLI tools is provided for most of the basic data mining
operations.  For more information on OpenSM or other Open Fabrics software see:
http://www.openfabrics.org/



ChangeLog
- Tue May 04 2010 Ira Weiny <weiny2@llnl.gov>

	* : 2.1.0-1 tag
	* : Update INSTALL doc

- Mon Apr 19 2010 Ira Weiny <weiny2@llnl.gov>

	* : Change output to match SKUMMEE's expectations on character
	limitations

- Thu Apr 15 2010 Ira Weiny <weiny2@llnl.gov>

	* : Made all error counters compatible with ibqueryerrors (libibmad)
	* : Added node-name-map support
	* : Added SKUMMEE output for all counters on all nodes.

- Wed Mar 02 2010 Ira Weiny <weiny2@llnl.gov>

	* : 2.0.0-1 tag

- Wed Aug 21 2008 Ira Weiny <weiny2@llnl.gov>

	* : Updated to new event plugin interface in master branch
	    WARNING: Now this tree is not compatible with opensm version <= 3.2.2
	* : Update Rcv Switch Relay errors message on SKUMMEE page

- Wed Jul 16 2008 Ira Weiny <weiny2@llnl.gov>

	* : Removed the glib dependancy.  malloc and free should be sufficient
	    for this operation.
	* : Added "Counter" variables to thresh DB.  These will give an
	    "impulse" type display to the variables.

- Mon May 12 2008 Ira Weiny <weiny2@llnl.gov>

	* : 1.0.1-4 tag
	* : Added opensmskummeequery man page
	* : Added README file
	* : updated INSTALL file to reflect the option of not using SKUMMEE and make
	    more generic

- Thu Mar 20 2008 Ira Weiny <weiny2@llnl.gov>

	* : 1.0.1-3 tag
	* : scripts/opensmskummee_update_thresh.pl.in: fix text for link down error

- Thu Mar 20 2008 Ira Weiny <weiny2@llnl.gov>

	* : 1.0.1-2 tag
	* : fix libopensmskummee.conf update error

- Thu Feb 14 2008 Ira Weiny <weiny2@llnl.gov>

	* : 1.0.1-1 tag
	* : Add option which allows for the suppression of errors in
	    opensmskummeequery output
	* : Add config option CLEAR_DB_ON_LOAD default = false

- Fri Dec 14 2007 Ira Weiny <weiny2@llnl.gov>

	* : Add genders support to be able to sumarize the fabric errors on
	    skummee

- Thu Dec 06 2007 Ira Weiny <weiny2@llnl.gov>

	* : Most functionality complete
	* : Add skummeeopensmquery to take care of allowing SKUMMEE to query the DB.
	    This CLI executable will also clear and report errors stored in the
		 plugin's DB.

- Thu Aug 30 2007 Ira Weiny <weiny2@llnl.gov>

	* : initial creation



src/sql_queries.c
/*
 * Copyright (C) 2007, Lawrence Livermore National Security, LLC.
 * Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 * Written by Ira Weiny weiny2@llnl.gov
 * CODE-235483
 * 
 * This file is part of libopensmskummeeplugin
 * For details, see http://www.llnl.gov/linux/
 *              or  http://www.nongnu.org/osmskummeeplugin/
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * OUR NOTICE AND TERMS AND CONDITIONS OF THE GNU GENERAL PUBLIC LICENSE
 *
 * Our Preamble Notice
 *
 * A. This notice is required to be provided under our contract with the U.S.
 * Department of Energy (DOE). This work was produced at the Lawrence Livermore
 * National Laboratory under Contract No.  DE-AC52-07NA27344 with the DOE.
 *
 */

#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <inttypes.h>

#include <mysql/mysql.h>
#include <mysql/mysqld_error.h>

#include "sql_queries.h"

/** =========================================================================
 * Define MYSQL query macros and functions to make code more readable.
 */
#define MYSQL_QUERY_FAIL() \
	plugin_log(plugin_data->osmlog, OSM_LOG_ERROR, "MYSQL Query failed : %s\n", query_str);

#define QUERY_STR_LEN (1024)
static char query_str[QUERY_STR_LEN];
int QUERY(plugin_data_t *plugin_data, const char *fmt, ...)
{
	int rc = 0;
	va_list args;
	va_start(args, fmt);
	vsnprintf(query_str, QUERY_STR_LEN, fmt, args);
	if ((rc = mysql_query(plugin_data->conn,query_str))) {
		MYSQL_QUERY_FAIL();
	}
	va_end(args);
	return (rc);
}

#define COL_TO_UINT64(num) \
	row[num] ? strtoull(row[num], NULL, 0) : 0;

/** =========================================================================
 * Ensure the node described by the port is in the nodes table
 */
static int
sql_add_node(plugin_data_t *plugin_data, osm_epi_port_id_t *port_id)
{
	MYSQL_RES *res;
	int ret = 0;

	if ((ret = QUERY(plugin_data, "select * from nodes where guid=%"PRIuLEAST64,
			port_id->node_guid))) {
		return (ret);
	}

	res = mysql_store_result(plugin_data->conn);
	if (mysql_num_rows(res) != 1) {
		/* the node is not in the table add it */
		ret = QUERY(plugin_data,
			"insert into nodes set guid=%"PRIuLEAST64",name=\"%s\";",
			port_id->node_guid,
			port_id->node_name);
	}

	mysql_free_result(res);

	return (ret);
}

static inline void
insert_port_errors(plugin_data_t *plugin_data, osm_epi_pe_event_t *pe_event)
{
	QUERY(plugin_data,
		"insert into port_errors set guid=%"PRIuLEAST64",port=%d,"
		"SymbolErrors=%"PRIuLEAST64","
		"LinkRecovers=%"PRIuLEAST64","
		"LinkDowned=%"PRIuLEAST64","
		"RcvErrors=%"PRIuLEAST64","
		"RcvRemotePhysErrors=%"PRIuLEAST64","
		"RcvSwitchRelayErrors=%"PRIuLEAST64","
		"XmtDiscards=%"PRIuLEAST64","
		"XmtConstraintErrors=%"PRIuLEAST64","
		"RcvConstraintErrors=%"PRIuLEAST64","
		"LinkIntegrityErrors=%"PRIuLEAST64","
		"ExcBufOverrunErrors=%"PRIuLEAST64","
		"VL15Dropped=%"PRIuLEAST64","
		"time_diff_s=%"PRIuLEAST64""
		";",
		pe_event->port_id.node_guid,
		pe_event->port_id.port_num,
		pe_event->symbol_err_cnt,
		pe_event->link_err_recover,
		pe_event->link_downed,
		pe_event->rcv_err,
		pe_event->rcv_rem_phys_err,
		pe_event->rcv_switch_relay_err,
		pe_event->xmit_discards,
		pe_event->xmit_constraint_err,
		pe_event->rcv_constraint_err,
		pe_event->link_integrity,
		pe_event->buffer_overrun,
		pe_event->vl15_dropped,
		pe_event->time_diff_s
		);
}

static inline void
update_port_errors(plugin_data_t *plugin_data, osm_epi_pe_event_t *pe_event)
{
	QUERY(plugin_data,
		"update port_errors set "
		"SymbolErrors=%"PRIuLEAST64","
		"LinkRecovers=%"PRIuLEAST64","
		"LinkDowned=%"PRIuLEAST64","
		"RcvErrors=%"PRIuLEAST64","
		"RcvRemotePhysErrors=%"PRIuLEAST64","
		"RcvSwitchRelayErrors=%"PRIuLEAST64","
		"XmtDiscards=%"PRIuLEAST64","
		"XmtConstraintErrors=%"PRIuLEAST64","
		"RcvConstraintErrors=%"PRIuLEAST64","
		"LinkIntegrityErrors=%"PRIuLEAST64","
		"ExcBufOverrunErrors=%"PRIuLEAST64","
		"VL15Dropped=%"PRIuLEAST64","
		"time_diff_s=%"PRIuLEAST64" "
		"where guid=%"PRIuLEAST64" and port=%d"
		";",
		pe_event->symbol_err_cnt,
		pe_event->link_err_recover,
		pe_event->link_downed,
		pe_event->rcv_err,
		pe_event->rcv_rem_phys_err,
		pe_event->rcv_switch_relay_err,
		pe_event->xmit_discards,
		pe_event->xmit_constraint_err,
		pe_event->rcv_constraint_err,
		pe_event->link_integrity,
		pe_event->buffer_overrun,
		pe_event->vl15_dropped,
		pe_event->time_diff_s,
		pe_event->port_id.node_guid,
		pe_event->port_id.port_num
		);
}

void
sql_add_port_errors(plugin_data_t *plugin_data, osm_epi_pe_event_t *pe_event)
{
	unsigned int num_fields;
	MYSQL_RES *res;
	MYSQL_ROW  row;
	osm_epi_port_id_t *port_id = &(pe_event->port_id);

	if (sql_add_node(plugin_data, port_id))
		return;

	if (QUERY(plugin_data,
			"select SymbolErrors, LinkRecovers,"
			"       LinkDowned,RcvErrors, RcvRemotePhysErrors,"
			"       RcvSwitchRelayErrors, XmtDiscards,"
			"       XmtConstraintErrors, RcvConstraintErrors,"
			"       LinkIntegrityErrors, ExcBufOverrunErrors,"
			"       VL15Dropped"
			"       from port_errors where guid=%"PRIuLEAST64" and port=%d",
			port_id->node_guid, port_id->port_num)) {
		return;
	}

	res = mysql_store_result(plugin_data->conn);

	if (mysql_num_rows(res) == 0) {
		/* simple insert for the first time */
		insert_port_errors(plugin_data, pe_event);
	} else if (mysql_num_rows(res) == 1) {
		/* must retrieve previous counts to be added to the new counts. */

		if ((num_fields = mysql_num_fields(res)) != 12) {
			plugin_log(plugin_data->osmlog, OSM_LOG_ERROR,
				"Failed to query port_error table for guid %"PRIuLEAST64", port %d; col ret %d != 12\n",
				port_id->node_guid, port_id->port_num, num_fields);
			goto free_res;
		}
		row = mysql_fetch_row(res);

		pe_event->symbol_err_cnt       += COL_TO_UINT64(0);
		pe_event->link_err_recover     += COL_TO_UINT64(1);
		pe_event->link_downed          += COL_TO_UINT64(2);
		pe_event->rcv_err              += COL_TO_UINT64(3);
		pe_event->rcv_rem_phys_err     += COL_TO_UINT64(4);
		pe_event->rcv_switch_relay_err += COL_TO_UINT64(5);
		pe_event->xmit_discards        += COL_TO_UINT64(6);
		pe_event->xmit_constraint_err  += COL_TO_UINT64(7);
		pe_event->rcv_constraint_err   += COL_TO_UINT64(8);
		pe_event->link_integrity       += COL_TO_UINT64(9);
		pe_event->buffer_overrun       += COL_TO_UINT64(10);
		pe_event->vl15_dropped         += COL_TO_UINT64(11);

		update_port_errors(plugin_data, pe_event);
	} else {
		/* more than one result? */
		plugin_log(plugin_data->osmlog, OSM_LOG_ERROR,
			"%llu port_error rows returned for guid %"PRIuLEAST64", port %d\n",
			mysql_num_rows(res), port_id->node_guid, port_id->port_num);
	}

free_res:
	mysql_free_result(res);
}

static inline void
insert_port_data_counters(plugin_data_t *plugin_data, osm_epi_dc_event_t *dc_event)
{
	QUERY(plugin_data,
		"insert into port_data_counters set guid=%"PRIuLEAST64",port=%d,"
		"xmit_data=%"PRIuLEAST64","
		"rcv_data=%"PRIuLEAST64","
		"xmit_pkts=%"PRIuLEAST64","
		"rcv_pkts=%"PRIuLEAST64","
		"unicast_xmit_pkts=%"PRIuLEAST64","
		"unicast_rcv_pkts=%"PRIuLEAST64","
		"multicast_xmit_pkts=%"PRIuLEAST64","
		"multicast_rcv_pkts=%"PRIuLEAST64","
		"time_diff_s=%"PRIuLEAST64""
		";",
		dc_event->port_id.node_guid, dc_event->port_id.port_num,
		dc_event->xmit_data,
		dc_event->rcv_data,
		dc_event->xmit_pkts,
		dc_event->rcv_pkts,
		dc_event->unicast_xmit_pkts,
		dc_event->unicast_rcv_pkts,
		dc_event->multicast_xmit_pkts,
		dc_event->multicast_rcv_pkts,
		dc_event->time_diff_s
		);
}

static inline void
update_port_data_counters(plugin_data_t *plugin_data, osm_epi_dc_event_t *dc_event)
{
	QUERY(plugin_data,
		"update port_data_counters set "
		"xmit_data=%"PRIuLEAST64","
		"rcv_data=%"PRIuLEAST64","
		"xmit_pkts=%"PRIuLEAST64","
		"rcv_pkts=%"PRIuLEAST64","
		"unicast_xmit_pkts=%"PRIuLEAST64","
		"unicast_rcv_pkts=%"PRIuLEAST64","
		"multicast_xmit_pkts=%"PRIuLEAST64","
		"multicast_rcv_pkts=%"PRIuLEAST64","
		"time_diff_s=%"PRIuLEAST64" "
		"where guid=%"PRIuLEAST64" and port=%d"
		";",
		dc_event->xmit_data,
		dc_event->rcv_data,
		dc_event->xmit_pkts,
		dc_event->rcv_pkts,
		dc_event->unicast_xmit_pkts,
		dc_event->unicast_rcv_pkts,
		dc_event->multicast_xmit_pkts,
		dc_event->multicast_rcv_pkts,
		dc_event->time_diff_s,
		dc_event->port_id.node_guid, dc_event->port_id.port_num
		);
}

void
sql_add_data_counters(plugin_data_t *plugin_data, osm_epi_dc_event_t *dc_event)
{
	unsigned int num_fields;
	MYSQL_RES *res;
	MYSQL_ROW  row;
	osm_epi_port_id_t *port_id = &(dc_event->port_id);

	if (sql_add_node(plugin_data, port_id))
		return;

	if (QUERY(plugin_data,
			"select xmit_data,rcv_data,xmit_pkts,rcv_pkts,"
			"       unicast_xmit_pkts,unicast_rcv_pkts,multicast_xmit_pkts,multicast_rcv_pkts"
			"       from port_data_counters where guid=%"PRIuLEAST64" and port=%d",
			port_id->node_guid, port_id->port_num)) {
		return;
	}

	res = mysql_store_result(plugin_data->conn);

	if (mysql_num_rows(res) == 0) {
		/* simple insert for the first time */
		insert_port_data_counters(plugin_data, dc_event);
	} else if (mysql_num_rows(res) == 1) {
		/* must retrieve previous counts to be added to the new counts. */

		if ((num_fields = mysql_num_fields(res)) != 8) {
			plugin_log(plugin_data->osmlog, OSM_LOG_ERROR,
				"Failed to query port_data_counters table for guid %"PRIuLEAST64", port %d; col ret %d != 8\n",
				port_id->node_guid, port_id->port_num, num_fields);
			goto free_res;
		}
		row = mysql_fetch_row(res);

		dc_event->xmit_data           += COL_TO_UINT64(0);
		dc_event->rcv_data            += COL_TO_UINT64(1);
		dc_event->xmit_pkts           += COL_TO_UINT64(2);
		dc_event->rcv_pkts            += COL_TO_UINT64(3);
		dc_event->unicast_xmit_pkts   += COL_TO_UINT64(4);
		dc_event->unicast_rcv_pkts    += COL_TO_UINT64(5);
		dc_event->multicast_xmit_pkts += COL_TO_UINT64(6);
		dc_event->multicast_rcv_pkts  += COL_TO_UINT64(7);

		update_port_data_counters(plugin_data, dc_event);
	} else {
		/* more than one result? */
		plugin_log(plugin_data->osmlog, OSM_LOG_ERROR,
			"%llu port_data_counters rows returned for guid %"PRIuLEAST64", port %d\n",
			mysql_num_rows(res), port_id->node_guid, port_id->port_num);
	}

free_res:
	mysql_free_result(res);
}

static void
insert_port_select(plugin_data_t *plugin_data, osm_epi_ps_event_t *ps_event)
{
	QUERY(plugin_data,
		"insert into port_select_counters set guid=%"PRIuLEAST64",port=%d,"
		"xmit_wait=%"PRIuLEAST64","
		"time_diff_s=%"PRIuLEAST64""
		";",
		ps_event->port_id.node_guid,
		ps_event->port_id.port_num,
		ps_event->xmit_wait,
		ps_event->time_diff_s
		);
}

static void
update_port_select(plugin_data_t *plugin_data, osm_epi_ps_event_t *ps_event)
{
	QUERY(plugin_data,
		"update port_select_counters set "
		"xmit_wait=%"PRIuLEAST64","
		"time_diff_s=%"PRIuLEAST64" "
		"where guid=%"PRIuLEAST64",port=%d,"
		";",
		ps_event->xmit_wait,
		ps_event->time_diff_s,
		ps_event->port_id.node_guid,
		ps_event->port_id.port_num
		);
}

void
sql_add_port_select(plugin_data_t *plugin_data, osm_epi_ps_event_t *ps_event)
{
	unsigned int num_fields;
	MYSQL_RES *res;
	MYSQL_ROW  row;
	osm_epi_port_id_t *port_id = &(ps_event->port_id);

	if (sql_add_node(plugin_data, port_id))
		return;

	if (QUERY(plugin_data,
			"select xmit_wait"
			"       from port_select_counters where guid=%"PRIuLEAST64" and port=%d",
			port_id->node_guid, port_id->port_num)) {
		return;
	}

	res = mysql_store_result(plugin_data->conn);

	if (mysql_num_rows(res) == 0) {
		/* simple insert for the first time */
		insert_port_select(plugin_data, ps_event);
	} else if (mysql_num_rows(res) == 1) {
		/* must retrieve previous counts to be added to the new counts. */

		if ((num_fields = mysql_num_fields(res)) != 1) {
			plugin_log(plugin_data->osmlog, OSM_LOG_ERROR,
				"Failed to query port_select_counters table for guid %"PRIuLEAST64", port %d; col ret %d != 1\n",
				port_id->node_guid, port_id->port_num, num_fields);
			goto free_res;
		}
		row = mysql_fetch_row(res);

		ps_event->xmit_wait += COL_TO_UINT64(0);

		update_port_select(plugin_data, ps_event);
	} else {
		/* more than one result? */
		plugin_log(plugin_data->osmlog, OSM_LOG_ERROR,
			"%llu port_select_counters rows returned for guid %"PRIuLEAST64", port %d\n",
			mysql_num_rows(res), port_id->node_guid, port_id->port_num);
	}

free_res:
	mysql_free_result(res);
}

int
sql_setup_db_conn(plugin_data_t *plugin_data)
{
	/* open DB connection */
	if (!mysql_init(&(plugin_data->mysql))) {
		plugin_log(plugin_data->osmlog, OSM_LOG_ERROR, "Failed to Connect to MySQL\n");
		return (1);
	}
	if (!(plugin_data->conn = mysql_real_connect(&(plugin_data->mysql), NULL,
			plugin_data->db_user, plugin_data->db_password,
			plugin_data->db_name, 0, NULL, 0))) {
		plugin_log(plugin_data->osmlog, OSM_LOG_ERROR, "Failed to open database \"%s\" : %s\n",
			plugin_data->db_name, mysql_error(&(plugin_data->mysql)));
		mysql_close(&(plugin_data->mysql));
		return (1);
	}


	if (plugin_data->clear_db_on_load) {
		/* Clean up the DB */
		QUERY(plugin_data, "delete from nodes;");
		QUERY(plugin_data, "delete from port_errors;");
		QUERY(plugin_data, "delete from port_data_counters;");
		QUERY(plugin_data, "delete from port_select_counters;");
	}

	plugin_log(plugin_data->osmlog, OSM_LOG_INFO,
		"Opened %s as user %s\n", plugin_data->db_name, plugin_data->db_user);
	return (0);
}



src/sql_queries.h
/*
 * Copyright (C) 2007, Lawrence Livermore National Security, LLC.
 * Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 * Written by Ira Weiny weiny2@llnl.gov
 * CODE-235483
 * 
 * This file is part of libopensmskummeeplugin
 * For details, see http://www.llnl.gov/linux/
 *              or  http://www.nongnu.org/osmskummeeplugin/
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * OUR NOTICE AND TERMS AND CONDITIONS OF THE GNU GENERAL PUBLIC LICENSE
 *
 * Our Preamble Notice
 *
 * A. This notice is required to be provided under our contract with the U.S.
 * Department of Energy (DOE). This work was produced at the Lawrence Livermore
 * National Laboratory under Contract No.  DE-AC52-07NA27344 with the DOE.
 *
 */

#ifndef _SQL_QUERIES_H_
#define _SQL_QUERIES_H_

#include "opensmskummeeplugin.h"

int  sql_setup_db_conn(plugin_data_t *plugin_data);
void sql_add_port_errors(plugin_data_t *plugin_data, osm_epi_pe_event_t *pe_event);
void sql_add_data_counters(plugin_data_t *plugin_data, osm_epi_dc_event_t *dc_event);
void sql_add_port_select(plugin_data_t *plugin_data, osm_epi_ps_event_t *ps_event);

#endif /* _SQL_QUERIES_H_ */



src/opensmskummeeplugin.c
/*
 * Copyright (C) 2007, Lawrence Livermore National Security, LLC.
 * Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 * Written by Ira Weiny weiny2@llnl.gov
 * CODE-235483
 * 
 * This file is part of libopensmskummeeplugin
 * For details, see http://www.llnl.gov/linux/
 *              or  http://www.nongnu.org/osmskummeeplugin/
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * OUR NOTICE AND TERMS AND CONDITIONS OF THE GNU GENERAL PUBLIC LICENSE
 *
 * Our Preamble Notice
 *
 * A. This notice is required to be provided under our contract with the U.S.
 * Department of Energy (DOE). This work was produced at the Lawrence Livermore
 * National Laboratory under Contract No.  DE-AC52-07NA27344 with the DOE.
 *
 */

#if HAVE_CONFIG_H
#  include <config.h>
#endif				/* HAVE_CONFIG_H */

#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>
#include <dlfcn.h>
#include <stdint.h>
#include <unistd.h>
#include <pthread.h>
#include <time.h>

#include <mysql/mysql.h>
#include <mysql/mysqld_error.h>

#include "opensmskummeeplugin.h"
#include "sql_queries.h"

/** =========================================================================
 * Called when new data comes in
 * This adds data to the FIFO
 */
static void
add_to_head(plugin_data_t *data, data_entry_t *entry)
{
	pthread_mutex_lock(&(data->lock));

	entry->next = data->head;
	entry->prev = NULL;

	if (data->head)
		data->head->prev = entry;

	data->head = entry;

	if (data->tail == NULL) {
		data->tail = entry;
	}
	pthread_mutex_unlock(&(data->lock));
}

/** =========================================================================
 * Remove and return the tail of the data.
 */
static data_entry_t *
pull_from_tail(plugin_data_t *data)
{
	data_entry_t *rc = NULL;
	pthread_mutex_lock(&(data->lock));
	if (!data->tail)
		goto out;

	if (data->tail->prev)
		data->tail->prev->next = NULL;

	rc = data->tail;
	data->tail = data->tail->prev;
	rc->prev = NULL;

	/* check if last element */
	if (rc == data->head)
		data->head = NULL;

out:
	pthread_mutex_unlock(&(data->lock));
	return (rc);
}

inline void
thread_signal(plugin_data_t *plugin_data)
{
	pthread_mutex_lock(&(plugin_data->sig_lock));
	pthread_cond_signal(&(plugin_data->signal));
	pthread_mutex_unlock(&(plugin_data->sig_lock));
}

/** =========================================================================
 * This is the main worker thread.  It looks for information comming in the
 * FIFO and submits the data to the database.  This ensures that the PerfMgr
 * (and therefore OpenSM threads) will not be delayed by DB activity.
 */
void *
db_write_thread(void *pd)
{
	plugin_data_t *plugin_data = (plugin_data_t *)pd;
	data_entry_t  *data = NULL;

	plugin_log(plugin_data->osmlog, OSM_LOG_INFO,
			"In DB write thread\n");

	plugin_data->exit_flag = 0;
	if (sql_setup_db_conn(plugin_data)) {
		thread_signal(plugin_data);
		return (NULL);
	}

	thread_signal(plugin_data);

	while (1) {
		if (plugin_data->exit_flag) {
			return (NULL);
		}

		if (!plugin_data->tail) {
			struct timespec ts;
			clock_gettime(CLOCK_REALTIME, &ts);
			ts.tv_sec += plugin_data->thread_sleep_s;
			pthread_mutex_lock(&(plugin_data->sig_lock));
			pthread_cond_timedwait(&(plugin_data->signal), &(plugin_data->sig_lock), &ts);
			pthread_mutex_unlock(&(plugin_data->sig_lock));
			continue;
		}

		data = pull_from_tail(plugin_data);

		if (!data) {
			plugin_log(plugin_data->osmlog, OSM_LOG_ERROR,
				"plugin_data->tail != NULL but pull returned NULL?");
			continue;
		}

		switch(data->type)
		{
			case OSM_EVENT_ID_PORT_ERRORS:
				sql_add_port_errors(plugin_data, &(data->data.pe_event));
				break;
			case OSM_EVENT_ID_PORT_DATA_COUNTERS:
				sql_add_data_counters(plugin_data, &(data->data.dc_event));
				break;
			case OSM_EVENT_ID_PORT_SELECT:
				sql_add_port_select(plugin_data, &(data->data.ps_event));
				break;
			case OSM_EVENT_ID_TRAP:
				break;
			case OSM_EVENT_ID_MAX: break;
		}

		free(data);
	}
}

/** =========================================================================
 */
static inline int
construct_plugin_data(plugin_data_t *data)
{
	FILE *fp = NULL;
	char  line[1024];
	char *last = NULL;
	char *key = NULL;
	char *val = NULL;

	if (pthread_mutex_init(&(data->lock), NULL)
			|| pthread_mutex_init(&(data->sig_lock), NULL)
			|| pthread_cond_init(&(data->signal), NULL)) {
		return (0);
	}

	data->db_user = DEF_DATABASE_USER;
	data->db_name = DEF_DATABASE_NAME;
	data->clear_db_on_load = 0;
	data->thread_sleep_s = DEF_THREAD_SLEEP;

	snprintf(line, 1023, "%s/%s", CONF_FILE_PREFIX, DATABASE_CONF);
	if (!(fp = fopen(line, "r"))) {
		return (0);
	}

	while (fgets(line, 1023, fp) != NULL) {
		/* get the first token */
		key = strtok_r(line, " \t\n", &last);
		if (!key)
			continue;
		val = strtok_r(NULL, " \t\n", &last);

		if (!strcmp("DB_USER", key) && val) {
			data->db_user = strdup(val);
		}
		if (!strcmp("DB_NAME", key) && val) {
			data->db_name = strdup(val);
		}
		if (!strcmp("DB_PASSWORD", key) && val) {
			data->db_password = strdup(val);
		}
		if (!strcmp("CLEAR_DB_ON_LOAD", key) && val) {
			data->clear_db_on_load = strcmp("FALSE", val);
		}
		if (!strcmp("THREAD_SLEEP_S", key) && val) {
			data->thread_sleep_s = abs((int)(strtol(val, NULL, 0)));
		}
	}
	fclose(fp);

	return (1);
}

static void
free_plugin_data(plugin_data_t *data)
{
	free(data->db_user);
	free(data->db_name);
	free(data->db_password);
	free(data);
}

/** =========================================================================
 */
static void *
create(struct osm_opensm *osm)
{
	struct timespec  ts;
	plugin_data_t   *plugin_data = NULL;
	pthread_attr_t   th_attr;
	int              rc = 0;

	if (!(plugin_data = malloc(sizeof(*plugin_data))))
		return (NULL);

	plugin_data->osmlog = &(osm->log);

	if (!construct_plugin_data(plugin_data)) {
		plugin_log(plugin_data->osmlog, OSM_LOG_ERROR, "Failed to read %s\n", DATABASE_CONF);
		free(plugin_data);
		return (NULL);
	}

	if (pthread_attr_init(&(th_attr))) {
		free_plugin_data(plugin_data);
		return (NULL);
	}
	pthread_mutex_lock(&(plugin_data->sig_lock));
	if (pthread_create(&(plugin_data->thread), &(th_attr),
              		db_write_thread, (void *)plugin_data)) {
		plugin_log(plugin_data->osmlog, OSM_LOG_INFO, "Failed to create DB write thread\n");
		pthread_attr_destroy(&(th_attr));
		free_plugin_data(plugin_data);
		return (NULL);
	}
	pthread_attr_destroy(&(th_attr));

	clock_gettime(CLOCK_REALTIME, &ts);
	ts.tv_sec += 2; /* give 2 sec to start up */
	rc = pthread_cond_timedwait(&(plugin_data->signal), &(plugin_data->sig_lock), &ts);
	pthread_mutex_unlock(&(plugin_data->sig_lock));
	if (rc == ETIMEDOUT) {
		plugin_log(plugin_data->osmlog, OSM_LOG_ERROR, "DB write thread failed to initialize\n");
		pthread_join(plugin_data->thread, NULL);
		free_plugin_data(plugin_data);
		return (NULL);
	}

	plugin_log(plugin_data->osmlog, OSM_LOG_INFO, "DB write thread started\n");

	return ((void *)plugin_data);
}

/** =========================================================================
 */
static void
delete(void *data)
{
	plugin_data_t *plugin_data = (plugin_data_t *)data;

	plugin_log(plugin_data->osmlog, OSM_LOG_INFO, "destroy called cleaning up...\n");

	plugin_data->exit_flag = 1;
	thread_signal(plugin_data);
	plugin_log(plugin_data->osmlog, OSM_LOG_INFO, "Stopping DB write thread...\n");
	pthread_join(plugin_data->thread, NULL);
	plugin_log(plugin_data->osmlog, OSM_LOG_INFO, "Closing DB connection...\n");
	mysql_close(&(plugin_data->mysql));
	free_plugin_data(plugin_data);
}

/** =========================================================================
 * All this function does is copy the data into our FIFO for later processing
 * by the worker thread.
 */
static void report(void *data, osm_epi_event_id_t event_id, void *event_data)
{
	plugin_data_t *plugin_data = (plugin_data_t *)data;
	data_entry_t *entry = (data_entry_t *)malloc(sizeof *entry);

	if (!entry) {
		plugin_log(plugin_data->osmlog, OSM_LOG_ERROR,
			"Reporting event \"%d\" : malloc failed : %s\n",
				event_id, strerror(errno));
		return;
	}

	entry->type = event_id;

	switch (event_id) {
	case OSM_EVENT_ID_PORT_ERRORS:
		memcpy (&(entry->data.pe_event), event_data, sizeof(entry->data.pe_event));
		break;
	case OSM_EVENT_ID_PORT_DATA_COUNTERS:
		memcpy (&(entry->data.dc_event), event_data, sizeof(entry->data.dc_event));
		break;
	case OSM_EVENT_ID_PORT_SELECT:
		memcpy (&(entry->data.ps_event), event_data, sizeof(entry->data.ps_event));
		break;
	case OSM_EVENT_ID_TRAP:
		memcpy (&(entry->data.trap_event), event_data, sizeof(entry->data.trap_event));
		break;
	case OSM_EVENT_ID_MAX:
	default:
		plugin_log(plugin_data->osmlog, OSM_LOG_ERROR,
			"Unknown event reported to plugin\n");
		free(entry);
		return;
	}

	add_to_head(data, entry);
}

/** =========================================================================
 * Define the object symbol for loading
 */
osm_event_plugin_t osm_event_plugin = {
      osm_version:OSM_VERSION,
      create:create,
      delete:delete,
      report:report
};
